File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.java
import java.io.*;
import java.util.*;
import java.util.Map.Entry;

public class Main {
	
	static int white, blue;
	static int[][] map;

	public static void main(String[] args) throws NumberFormatException, IOException {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st;
		
		int N = Integer.parseInt(br.readLine());
		map = new int[N][N];
		for(int i=0; i<N; i++) {
			st = new StringTokenizer(br.readLine());
			for(int j=0; j<N; j++) {
				map[i][j] = Integer.parseInt(st.nextToken());
			}
		}
	
		DQ(N, 0, 0);
		
		sb.append(white).append('\n');
		sb.append(blue).append('\n');
		System.out.println(sb);
	}
	
	private static void DQ(int N, int startR, int startC) {
		if(N == 0) return;
		
		if(check(N, startR, startC, 0) || check(N, startR, startC, 1)) {
			return;
		}
		
		DQ(N/2, startR, startC);
		DQ(N/2, startR, startC+N/2);
		DQ(N/2, startR+N/2, startC);
		DQ(N/2, startR+N/2, startC+N/2);
		
	}
	
	private static boolean check(int N, int startR, int startC, int num) {
		for(int i=startR; i<startR+N; i++) {
			for(int j=startC; j<startC+N; j++) {
				if(map[i][j] != num) {
					return false;
				}
			}
		}
		
		if(num == 1) blue++;
		else white++;
		return true;
	}
}

File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.cc
#define _CRT_SECURE_NO_WARNINGS
#include <iostream>
#include <vector>
#include <algorithm>
#include <math.h>
#include <queue>
using namespace std;

static int white, blue;
static int map[128][128];

bool check(int N, int R, int C, int num) {
	for (int i = R; i < R + N; i++) {
		for (int j = C; j < C + N; j++) {
			if (map[i][j] != num)
				return false;
		}
	}
	if (num == 1) blue++;
	else white++;

	return true;
}

void DQ(int N, int R, int C) {
	if (N == 0 || check(N, R, C, 0) || check(N, R, C, 1))
		return;

	DQ(N / 2, R, C);
	DQ(N / 2, R + N / 2, C);
	DQ(N / 2, R, C + N / 2);
	DQ(N / 2, R + N / 2, C + N / 2);
	return;
}


int main(int argc, char* argv[]) {

	int N;
	scanf("%d", &N);
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++) {
			scanf("%1d", &map[i][j]);
		}
	}

	DQ(N, 0, 0);

	printf("%d\n%d\n", white, blue);
	
	return 0;
}